### PR TITLE
List Connected devices in bridged AP mode

### PR DESCRIPTION
**Pull request by Taikuh**
For #538 

Because the upstream router takes care of DHCP requests in bridged mode, `arp` doesn't cache any entries. We use `iw dev wlan0 station dump` to at least find the MAC addresses of devices connected to wlan0.

A hacky way to find the IP addresses of these connected devices is to do an IP scan of the LAN with `nmap`. Any pinged device from this scan will have its data cached by arp (EXCEPT for the hostname!!). However, this adds another dependency to RaspAP.

## What I've Done
- `$clients` in bridged mode is a list of connected devices MAC addrs
- `$clients` in default mode still functions the same, but I simplified the exec() internals
- info text appears in bridged mode: tells users to see their own router page for more info
- "More" link directs to hostapd page rather than dhcp for bridged mode.

## How to proceed
1. Should I use nmap to at least get IP of connected devices in bridged mode?
2. In default routed mode, should Connected Devices use iw station dump to further filter out unconnected devices that still have a dnsmasq lease?

### CODE REVIEW

**Comment by billz** [comment]
@Taikuh you are on fire, good sir 🧯
approach is sound, will do a pull and check it out. thanks :)

File: templates/dashboard.php [hunk 1]
```diff
@@ -1,11 +1,18 @@
...
-    $client_iface = RASPI_WIFI_CLIENT_INTERFACE;
+  $client_iface = RASPI_WIFI_CLIENT_INTERFACE;
+}
+$MACPattern = '"([[:xdigit:]]{2}:){5}[[:xdigit:]]{2}"';
+if ($arrHostapdConf['BridgedEnable'] == 1) {
+  $moreLink = "index.php?page=hostapd_conf";
+  exec('iw dev '.$client_iface.' station dump | grep -oE '.$MACPattern, $clients);
+} else {
+  $moreLink = "index.php?page=dhcpd_conf";
+  exec('cat '.RASPI_DNSMASQ_LEASES.'| grep -E $(arp -i '.$client_iface.' -n | grep -oE '.$MACPattern.' | paste -sd "|")', $clients);
```
Comment (Taikuh) [inline templates/dashboard.php]:
```suggestion
  exec('cat '.RASPI_DNSMASQ_LEASES.'| grep -E $(iw dev '.$client_iface.' station dump | grep -oE '.$MACPattern.' | paste -sd "|")', $clients);
```

iw station dump is more up-to-date than arp cache

**Comment by billz** [comment]
> Should I use nmap to at least get IP of connected devices in bridged mode?

With the notice indicating bridge mode think it's acceptable to display just the mac. We can guess how some users might react to seeing nmap being added by the installer ;)

> In default routed mode, should Connected Devices use iw station dump to further filter out unconnected devices that still have a dnsmasq lease?

Yep, makes sense. Nice touch with the 'more' link, btw.

I think the small 'text-muted' class works well here:

Untitled

File: templates/dashboard.php [hunk 2]
```diff
@@ -71,25 +78,38 @@
...
+                          <th><?php echo _("Host name"); ?></th>
+                          <th><?php echo _("IP Address"); ?></th>
+                          <th><?php echo _("MAC Address"); ?></th>
+                        <?php endif; ?>
                       </tr>
                     </thead>
                     <tbody>
+                        <?php if ($arrHostapdConf['BridgedEnable'] == 1) : ?>
+                          <tr>
+                            <td><?php echo _("<em>Bridged AP mode is enabled. For Hostname and IP, see your router's admin page.</em>");?></td>
```
Comment (Taikuh) [inline templates/dashboard.php]:
```suggestion
                            <td><small class="text-muted"><?php echo _("Bridged AP mode is enabled. For Hostname and IP, see your router's admin page.");?></small></td>
```

Good call, I should familiarize myself more with bootstrap

File: locale/en_US/LC_MESSAGES/messages.po [hunk 3]
```diff
@@ -663,6 +663,10 @@ msgstr "Attempting to start TOR"
 msgid "Attempting to stop TOR"
 msgstr "Attempting to stop TOR"
 
+#: template/dashboard.php
+msgid "<em>Bridged AP mode is enabled. For Hostname and IP, see your router's admin page.</em>"
```
Comment (Taikuh) [inline locale/en_US/LC_MESSAGES/messages.po]:
```suggestion
msgid "Bridged AP mode is enabled. For Hostname and IP, see your router's admin page."
```

File: locale/en_US/LC_MESSAGES/messages.po [hunk 4]
```diff
@@ -663,6 +663,10 @@ msgstr "Attempting to start TOR"
...
+msgstr "<em>Bridged AP mode is enabled. For Hostname and IP, see your router's admin page.</em>"
```
Comment (Taikuh) [inline locale/en_US/LC_MESSAGES/messages.po]:
```suggestion
msgstr "Bridged AP mode is enabled. For Hostname and IP, see your router's admin page."
```

**Comment by Taikuh** [comment]
> We can guess how some users might react to seeing nmap being added by the installer ;)

Heh, fill me in, what rep does `nmap` have in this world?

> Nice touch with the 'more' link, btw.

I'd say next step would be to add a **Connected Devices** tab under **Hotspot** and to link directly to it.  
The default routed mode's More link could even direct to that tab too.  
Pro: more consistent UX.  
Con: maybe users prefer a shortcut to viewing their IP leases?

**Comment by billz** [comment]
> what rep does nmap have in this world?

probably none. a bit much to install a new package just to resolve client IPs though. maybe if it becomes useful elsewhere in the project

> next step would be to add a Connected Devices tab under Hotspot and to link directly to it

agreed, for UX consistency. this differs enough from dhcp leases for it to make sense. also aligns with the DHCP server admin.

Nice addition overall. thanks for the quick turnaround! 

Resolves #538

**Pull request merged**